### PR TITLE
NOISSUE - Implementation of gRPC mTLS

### PR DESCRIPTION
**Pull request by arvindh123**
Pull request title should be `MF-XXX - description` or `NOISSUE - description` where XXX is ID of issue that this PR relate to.
Please review the [CONTRIBUTING.md](https://github.com/mainflux/mainflux/blob/master/CONTRIBUTING.md) file for detailed contributing guidelines.

### What does this do?
This PR aims to add MTLS (Mutual Transport Layer Security) support for the gRPC client in the Mainflux . It includes changes to both the gRPC server side and client side. The implementation of MTLS ensures secure communication between the client and server by requiring authentication and encryption for the transmitted data. 

### Which issue(s) does this PR fix/relate to?
Put here `Resolves #XXX` to auto-close the issue that your PR fixes (if such)
No

### List any changes that modify/break current functionality
No

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -120,4 +127,10 @@ rundev:
...
 	sed -i "s,file: brokers/.*.yml,file: brokers/${MF_BROKER_TYPE}.yml," docker/docker-compose.yml
 	sed -i "s,MF_BROKER_URL=.*,MF_BROKER_URL=$$\{MF_$(shell echo ${MF_BROKER_TYPE} | tr 'a-z' 'A-Z')_URL\}," docker/.env
-	docker-compose -f docker/docker-compose.yml up
+	docker-compose -f docker/docker-compose.yml -p $(DOCKER_PROJECT) up
+
+run_mtls_grpc:
+	sed -i "s,file: brokers/.*.yml,file: brokers/${MF_BROKER_TYPE}.yml," docker/docker-compose.yml
+	sed -i "s,MF_BROKER_URL=.*,MF_BROKER_URL=$$\{MF_$(shell echo ${MF_BROKER_TYPE} | tr 'a-z' 'A-Z')_URL\}," docker/.env
+	make -C docker/ssl users_grpc_certs things_grpc_certs
+	docker-compose -f docker/docker-compose.yml -f docker/ssl/docker-compose.grpc-mtls.yaml -p $(DOCKER_PROJECT) up
```
Comment (dborovcanin) [inline Makefile]: Add an empty line.

File: docker/addons/bootstrap/docker-compose.grpc-mtls.yaml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+      MF_AUTH_GRPC_CLIENT_MTLS: ${MF_USERS_GRPC_CLIENT_MTLS}
+      MF_AUTH_GRPC_CLIENT_TLS: ${MF_USERS_GRPC_CLIENT_TLS}
+      MF_AUTH_GRPC_CLIENT_CERT: /users-grpc-client.crt
+      MF_AUTH_GRPC_CLIENT_KEY: /users-grpc-client.key
+      MF_AUTH_GRPC_SERVER_CA_CERTS: /users-grpc-server-ca.crt
+    volumes:
+      # Users gRPC client certificates
+      - ${MF_USERS_GRPC_CLIENT_CERT}:/users-grpc-client.crt
+      - ${MF_USERS_GRPC_CLIENT_KEY}:/users-grpc-client.key
+      - ${MF_USERS_GRPC_SERVER_CA_CERTS}:/users-grpc-server-ca.crt
```
Comment (dborovcanin) [inline docker/addons/bootstrap/docker-compose.grpc-mtls.yaml]: Add an empty line.

File: docker/addons/cassandra-reader/docker-compose.grpc-mtls.yaml [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+      MF_THINGS_AUTH_GRPC_SERVER_CA_CERTS: /server_ca.crt
+    volumes:
+      # Users gRPC client certificates
+      - ${MF_USERS_GRPC_CLIENT_CERT}:/users-grpc-client.crt
+      - ${MF_USERS_GRPC_CLIENT_KEY}:/users-grpc-client.key
+      - ${MF_USERS_GRPC_SERVER_CA_CERTS}:/users-grpc-server-ca.crt
+      # Things gRPC client certificates
+      - ${MF_THINGS_AUTH_GRPC_CLIENT_CERT}:/client.crt
+      - ${MF_THINGS_AUTH_GRPC_CLIENT_KEY}:/client.key
+      - ${MF_THINGS_AUTH_GRPC_SERVER_CA_CERTS}:/server_ca.crt
```
Comment (dborovcanin) [inline docker/addons/cassandra-reader/docker-compose.grpc-mtls.yaml]: Same.

Comment (dborovcanin) [reply docker/addons/cassandra-reader/docker-compose.grpc-mtls.yaml]: Check all the other cases.

File: docker/ssl/docker-compose.grpc-mtls.yaml [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+      MF_THINGS_AUTH_GRPC_CLIENT_CERT: /client.crt
+      MF_THINGS_AUTH_GRPC_CLIENT_KEY: /client.key
+      MF_THINGS_AUTH_GRPC_SERVER_CA_CERTS: /server_ca.crt
+    volumes:
+      # Things gRPC client certificates
+      - ${MF_THINGS_AUTH_GRPC_CLIENT_CERT}:/client.crt
+      - ${MF_THINGS_AUTH_GRPC_CLIENT_KEY}:/client.key
+      - ${MF_THINGS_AUTH_GRPC_SERVER_CA_CERTS}:/server_ca.crt
+
+  # bootstrap:
```
Comment (dborovcanin) [inline docker/ssl/docker-compose.grpc-mtls.yaml]: Remove commented code.

File: internal/clients/grpc/connect.go [hunk 5]
```diff
@@ -19,10 +23,13 @@ var (
 )
 
 type Config struct {
-	ClientTLS bool          `env:"CLIENT_TLS"    envDefault:"false"`
-	CACerts   string        `env:"CA_CERTS"      envDefault:""`
-	URL       string        `env:"URL"           envDefault:""`
-	Timeout   time.Duration `env:"TIMEOUT"       envDefault:"1s"`
+	ClientTLS    bool          `env:"CLIENT_TLS"       envDefault:"false"`
+	ClientMTLS   bool          `env:"CLIENT_MTLS"      envDefault:"false"`
```
Comment (dborovcanin) [inline internal/clients/grpc/connect.go]: Do we need these flags? Is it going to be simpler if we decide this dynamically: if the cert is passed, we use TLS, if the client cert or key is passed, we go for mTLS.

File: internal/clients/grpc/connect.go [hunk 6]
```diff
@@ -105,3 +144,14 @@ func (c *Client) Secure() string {
 	}
 	return "without TLS"
 }
+
+func loadCertFile(certFile string) ([]byte, error) {
```
Comment (dborovcanin) [inline internal/clients/grpc/connect.go]: This method "smells", i.e. returning `nil, nil` by default does not look safe.

**Review by dborovcanin — CHANGES_REQUESTED** [review]
We need documentation for this, probably with a PR to [MF docs](https://github.com/mainflux/docs/pulls).

File: internal/clients/grpc/connect.go (hunk 6 above)
Comment (arvindh123) [reply internal/clients/grpc/connect.go]: Fixed, Please check and give your feedback

File: Makefile (hunk 1 above)
Comment (arvindh123) [reply Makefile]: Added

File: docker/addons/bootstrap/docker-compose.grpc-mtls.yaml (hunk 2 above)
Comment (arvindh123) [reply docker/addons/bootstrap/docker-compose.grpc-mtls.yaml]: Added

File: docker/ssl/docker-compose.grpc-mtls.yaml (hunk 4 above)
Comment (arvindh123) [reply docker/ssl/docker-compose.grpc-mtls.yaml]: Removed

File: Makefile [hunk 7]
```diff
@@ -54,9 +61,30 @@ define make_docker_dev
...
+ifneq ($(filter run_addons%,$(firstword $(MAKECMDGOALS))),)
+  temp_args := $(wordlist 2,$(words $(MAKECMDGOALS)),$(MAKECMDGOALS))
+  RUN_ADDON_ARGS :=  $(filter-out up down,$(if $(temp_args),$(temp_args),$(ADDON_SERVICES) $(EXTERNAL_SERVICES)))
+  $(eval $(RUN_ADDON_ARGS):;@)
+endif
+
+FILTERED_SERVICE = $(filter-out $(RUN_ADDON_ARGS), $(SERVICES))
 
-.PHONY: all $(SERVICES) dockers dockers_dev latest release
+all: $(SERVICE)
```
Comment (rodneyosodo) [inline Makefile]:
```suggestion
all: $(SERVICES)
```

**Review by rodneyosodo — CHANGES_REQUESTED** [review]
I am getting these warning on 
```bash
make run args "-d"


GNU Make 4.3
Built for x86_64-pc-linux-gnu
Copyright (C) 1988-2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Reading makefiles...
Reading makefile 'Makefile'...
Updating makefiles....
 Considering target file 'Makefile'.
  Looking for an implicit rule for 'Makefile'.
  Trying pattern rule with stem 'Makefile'.
  Trying implicit prerequisite 'Makefile.o'.
  Trying pattern rule with stem 'Makefile'.
  Trying implicit prerequisite 'Makefile.c'.
  Trying pattern rule with stem 'Makefile'.
  Trying implicit prerequisite 'Makefile.cc'.
  Trying pattern rule with stem 'Makefile'.
  Trying implicit prerequisite 'Makefile.C'.
  Trying pattern rule with stem 'Makefile'.
  Trying implicit prerequisite 'Makefile.cpp'.
```

File: Makefile (hunk 7 above)
Comment (arvindh123) [reply Makefile]: Correct command , `args` is a make variable 
```bash 
make run args="-d"
```

File: docker/addons/bootstrap/docker-compose.grpc-mtls.yaml [hunk 8]
```diff
@@ -0,0 +1,14 @@
+services:
+  bootstrap:
+    environment:
+      # Users gRPC client environmental varaibles
+      MF_AUTH_GRPC_CLIENT_MTLS: ${MF_USERS_GRPC_CLIENT_MTLS}
+      MF_AUTH_GRPC_CLIENT_TLS: ${MF_USERS_GRPC_CLIENT_TLS}
+      MF_AUTH_GRPC_CLIENT_CERT: /users-grpc-client.crt
```
Comment (rodneyosodo) [inline docker/addons/bootstrap/docker-compose.grpc-mtls.yaml]:
```suggestion
      MF_AUTH_GRPC_CLIENT_CERT: ${MF_AUTH_GRPC_CLIENT_CERT}
```

Then we update `.env` file same way we do with vault token

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: docker/addons/bootstrap/docker-compose.grpc-mtls.yaml (hunk 8 above)
Comment (arvindh123) [reply docker/addons/bootstrap/docker-compose.grpc-mtls.yaml]: @rodneyosodo .  The `MF_AUTH_GRPC_CLIENT_CERT` variables are shared with both addons and mainflux core service , 
For mainflux core service path will be `./ssl/certs` for addons path will be `../../ssl/certs` 
So only added `ADDON_CERT_PREFIX_PATH` variable for addon services

File: docker/.env [hunk 9]
```diff
@@ -402,9 +422,12 @@ MF_SMPP_NOTIFIER_INSTANCE_ID=
 
 ### GRAFANA and PROMETHEUS
 MF_PROMETHEUS_PORT=9090
-MF_GRAFANA_PORT=3000
+MF_GRAFANA_PORT=5001
```
Comment (rodneyosodo) [inline docker/.env]:
```suggestion
MF_GRAFANA_PORT=3000
```

File: docker/.env [hunk 10]
```diff
@@ -61,6 +62,15 @@ MF_USERS_RESET_PWD_TEMPLATE=users.tmpl
 MF_USERS_PASS_REGEX=^.{8,}$$
 MF_USERS_INSTANCE_ID=
 
+### Users gRPC Client
+MF_USERS_GRPC_URL=users:7001
+MF_USERS_GRPC_TIMEOUT=1s
+MF_USERS_GRPC_CLIENT_TLS=true
```
Comment (rodneyosodo) [inline docker/.env]: Do we need this `*_GRPC_CLIENT_TLS` and `*_GRPC_CLIENT_MTLS` since they aren't used

File: internal/clients/grpc/connect.go [hunk 11]
```diff
@@ -45,20 +50,52 @@ func NewClientHandler(c *Client) ClientHandler {
...
 		if err != nil {
-			return nil, secure, err
+			fmt.Printf("failed to load system certificate %s\n", err.Error())
 		}
+		if sysCertPool != nil {
+			tlsConfig.RootCAs = sysCertPool
+		}
+
+		// Loading root ca certificates file
+		rootCA, err := loadCertFile(cfg.ServerCAFile)
```
Comment (rodneyosodo) [inline internal/clients/grpc/connect.go]: Since we check `if cfg.ServerCAFile != ""` on L59 the `cfg.ServerCAFile` can't be empty on L72 and hence L148. I'd suggest to remove `loadCertFile` and have something like 
```go
		// Loading root ca certificates file
		rootCA, err := ioutil.ReadFile(cfg.ServerCAFile)
		if err != nil {
			return nil, secure, fmt.Errorf("failed to read root ca file: %w", err)
		}
```

File: internal/clients/grpc/connect.go [hunk 12]
```diff
@@ -45,20 +50,52 @@ func NewClientHandler(c *Client) ClientHandler {
...
 	tc := insecure.NewCredentials()
 
-	if cfg.ClientTLS && cfg.CACerts != "" {
-		var err error
-		tc, err = credentials.NewClientTLSFromFile(cfg.CACerts, "")
+	if cfg.ServerCAFile != "" {
+		tlsConfig := &tls.Config{}
+
+		// Loading System certificates
+		sysCertPool, err := x509.SystemCertPool()
```
Comment (rodneyosodo) [inline internal/clients/grpc/connect.go]: can `x509.SystemCertPool()` return an empty `sysCertPool` together with a `nil` error? If not I suggest this refactor

```go
		// Loading System certificates
		sysCertPool, err := x509.SystemCertPool()
		if err != nil {
			return nil, secure, fmt.Errorf("failed to load system cert pool: %w", err)
		}
		tlsConfig.RootCAs = sysCertPool
```

File: internal/server/grpc/grpc.go [hunk 13]
```diff
@@ -101,3 +150,14 @@ func (s *Server) Stop() error {
 
 	return nil
 }
+
+func loadCertFile(certFile string) ([]byte, error) {
```
Comment (rodneyosodo) [inline internal/server/grpc/grpc.go:151]: Suggesting

```go
func loadCertFile(certFile string) ([]byte, error) {
	if certFile == "" {
		return nil, fmt.Errorf("cert file is empty")
	}

	return os.ReadFile(certFile)
}

```

File: internal/server/grpc/grpc.go [hunk 14]
```diff
@@ -47,30 +51,75 @@ func New(ctx context.Context, cancel context.CancelFunc, name string, config ser
...
+	}
 
 	listener, err := net.Listen("tcp", s.Address)
 	if err != nil {
 		return fmt.Errorf("failed to listen on port %s: %w", s.Address, err)
 	}
+	creds := grpc.Creds(insecure.NewCredentials())
 
 	switch {
 	case s.Config.CertFile != "" || s.Config.KeyFile != "":
```
Comment (rodneyosodo) [inline internal/server/grpc/grpc.go:65]:
```suggestion
	case s.Config.CertFile != "" && s.Config.KeyFile != "":
```

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: internal/clients/grpc/connect.go (hunk 5 above)
Comment (arvindh123) [reply internal/clients/grpc/connect.go]: Removed flags

File: Makefile (hunk 7 above)
Comment (arvindh123) [reply Makefile]: Done changed to `SERVICES`

File: docker/.env (hunk 9 above)
Comment (arvindh123) [reply docker/.env]: Will change back to original

Comment (arvindh123) [reply docker/.env]: Done

File: docker/.env (hunk 10 above)
Comment (arvindh123) [reply docker/.env]: Will remove it , Planned to remove in last commit,

File: internal/server/grpc/grpc.go (hunk 13 above)
Comment (arvindh123) [reply internal/server/grpc/grpc.go:151]: @rodneyosodo , 
In server, if the CertFile is not provided, then it should automatically switch to non-tls http
So if there is no file it should ignore and run http server 
To make the function better , I have modified like as given below
```go
func loadCertFile(certFile string) ([]byte, error) {
	var dd []byte
	if certFile != "" {
		var err error
		dd, err = ioutil.ReadFile(certFile)
		if err != nil {
			return nil, err
		}
	}
	return dd, nil
}
```

File: internal/server/grpc/grpc.go (hunk 14 above)
Comment (arvindh123) [reply internal/server/grpc/grpc.go:65]: @rodneyosodo 
Since there is no variable to check weather user is opted from TLS , 
So if any one of the file is given then assumes user is opted for TLS.

If user fails to provide two file, Instead provided one file, Then program will run without TLS 
So in order to notify the uesr, if one file is provided it will lead to error , So by user will get notified

File: internal/clients/grpc/connect.go (hunk 12 above)
Comment (arvindh123) [reply internal/clients/grpc/connect.go]: @rodneyosodo , System Certs is removed in this PR, It will be added in separate   PR

File: internal/clients/grpc/connect.go (hunk 11 above)
Comment (arvindh123) [reply internal/clients/grpc/connect.go]: Done

File: docker/.env (hunk 10 above)
Comment (arvindh123) [reply docker/.env]: Removed env variable  from env and docker-compose.

File: internal/server/grpc/grpc.go (hunk 13 above)
Comment (rodneyosodo) [reply internal/server/grpc/grpc.go:151]: What about something like this
```go
func loadCertFile(certFile string) ([]byte, error) {
	var dd []byte
	if certFile != "" {
		return os.ReadFile(certFile)
	}
	return dd, nil
}
```

File: internal/server/grpc/grpc.go (hunk 14 above)
Comment (rodneyosodo) [reply internal/server/grpc/grpc.go:65]: Understood

File: internal/clients/grpc/connect.go [hunk 15]
```diff
@@ -45,20 +50,42 @@ func NewClientHandler(c *Client) ClientHandler {
...
 	tc := insecure.NewCredentials()
 
-	if cfg.ClientTLS && cfg.CACerts != "" {
-		var err error
-		tc, err = credentials.NewClientTLSFromFile(cfg.CACerts, "")
+	if cfg.ServerCAFile != "" {
+		tlsConfig := &tls.Config{}
+
+		// Loading root ca certificates file
+		rootCA, err := ioutil.ReadFile(cfg.ServerCAFile)
```
Comment (rodneyosodo) [inline internal/clients/grpc/connect.go]: `ioutil.ReadFile` is deprecated as of 1.16

```suggestion
		rootCA, err := os.ReadFile(cfg.ServerCAFile)
```

File: internal/clients/grpc/connect.go [hunk 16]
```diff
@@ -45,20 +50,42 @@ func NewClientHandler(c *Client) ClientHandler {
...
+		if rootCA != nil {
+			capool := x509.NewCertPool()
+			if !capool.AppendCertsFromPEM(rootCA) {
+				return nil, secure, fmt.Errorf("failed to append root ca to tls.Config")
+			}
+			tlsConfig.RootCAs = capool
 		}
+
+		// Loading mtls certificates file
+		if cfg.ClientCert != "" || cfg.ClientKey != "" {
```
Comment (rodneyosodo) [inline internal/clients/grpc/connect.go:85]: Since `tls.LoadX509KeyPair` will return an error if either is empty

```suggestion
		if cfg.ClientCert != "" && cfg.ClientKey != "" {
```

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: internal/clients/grpc/connect.go (hunk 16 above)
Comment (arvindh123) [reply internal/clients/grpc/connect.go:85]: @rodneyosodo  As said before , the same reason, 

Since there is no variable to check weather user is opted from HTTPS ,

If user fails to provide two file, Instead provided one file, Then program will run without TLS if we use  `&&`

Instead if we use `||` if one file is provided it will lead to error , So by user will get notified

Comment (rodneyosodo) [reply internal/clients/grpc/connect.go:85]: Okay. I see.

File: docker/addons/mongodb-reader/docker-compose.grpc-mtls.yaml [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+      MF_AUTH_GRPC_CLIENT_CERT: /users-grpc-client.crt
+      MF_AUTH_GRPC_CLIENT_KEY: /users-grpc-client.key
+      MF_AUTH_GRPC_SERVER_CA_CERTS: /users-grpc-server-ca.crt
+      # Things gRPC client environmental varaibles
+      MF_THINGS_AUTH_GRPC_CLIENT_CERT: /client.crt
+      MF_THINGS_AUTH_GRPC_CLIENT_KEY: /client.key
+      MF_THINGS_AUTH_GRPC_SERVER_CA_CERTS: /server_ca.crt
+    volumes:
+      # Users gRPC client certificates
+      - ${MF_ADDONS_CERTS_PATH_PREFIX}${MF_USERS_GRPC_CLIENT_CERT}:/users-grpc-client.crt
```
Comment (mteodor) [inline docker/addons/mongodb-reader/docker-compose.grpc-mtls.yaml]: why two envs and then a constant suffix
wouldnt be easier to have only MF_USERS_GRPC_CLIENT_CERT to specify the full path to file
also should these variables be added to README.md?

File: internal/clients/grpc/connect.go [hunk 18]
```diff
@@ -45,20 +50,42 @@ func NewClientHandler(c *Client) ClientHandler {
...
+			}
+			tlsConfig.Certificates = []tls.Certificate{certificate}
+		}
+
+		tc = credentials.NewTLS(tlsConfig)
 		secure = true
 	}
 
-	opts = append(opts, gogrpc.WithTransportCredentials(tc), gogrpc.WithUnaryInterceptor(otelgrpc.UnaryClientInterceptor()))
+	opts = append(opts, gogrpc.WithTransportCredentials(tc))
```
Comment (mteodor) [inline internal/clients/grpc/connect.go:97]: why is this removed `gogrpc.WithUnaryInterceptor(otelgrpc.UnaryClientInterceptor()`?

Comment (mteodor) [reply internal/clients/grpc/connect.go:97]: i see added before

File: docker/addons/mongodb-reader/docker-compose.grpc-mtls.yaml (hunk 17 above)
Comment (arvindh123) [reply docker/addons/mongodb-reader/docker-compose.grpc-mtls.yaml]: @mteodor 
Env Var `MF_USERS_GRPC_CLIENT_CERT` is commonly used by all the service 
From `docker/docker-compose.yml`, path in present in  `MF_USERS_GRPC_CLIENT_CERT` will be valid , but 
from `docker/addons/docker-compose.yml` path present in ` MF_USERS_GRPC_CLIENT_CERT` will be invalid. 
So make the path valid for `addons` a prefix var `MF_ADDONS_CERTS_PATH_PREFIX` is used

**Review by rodneyosodo — APPROVED** [review]
LGTM

**Review by drasko — COMMENTED** [review]
@dborovcanin to review, please

**Review by dborovcanin — CHANGES_REQUESTED** [review]
Overall, the PR looks good. However, Makefile and the whole process of running services looks a little bit to complex now. I'd like you try to:

1. Avoid repetition of compose files (especially since ssl/docker-compose.grpc-mtls.yml is just a repetition of 2 sections)
2. Avoid two compose files for add-ons, as well
3. Avoid special prefixes for addons
4. Simplify Makefile (`ssl/Makefile` looks fine, only the root Makefile needs some love)

**Comment by arvindh123** [comment]
@mainflux/contributors  @mainflux/maintainers 
To test this PR and to run gRPC with mTLS , following docs will be helpful 
## Running Mainflux Core Services

To run all the Mainflux core services, use the following command:

```
make run
```

This command will start all the mainflux core services.

If you want to run the Mainflux core services with gRPC mTLS enabled, use the following command:

```
GRPC_MTLS=true make run
```

If you want to run the Mainflux core services with gRPC TLS enabled, use the following command:

```
GRPC_TLS=true make run
```

This will start all the mainflux core services with secure communication using automatically generated certificates if they are not already available.

## Running Specific Addons

If you only want to run specific addons, you can specify them in the command. For example, to run the `timescale-reader` and `timescale-writer` addons, use the following command:

```
make run_addons timescale-reader timescale-writer
```

This command will start the specified addons.

If you want to run specific addons with gRPC mTLS enabled, use the following command:

```
GRPC_MTLS=true make run_addons timescale-reader timescale-writer
```

If you want to run specific addons with gRPC TLS enabled, use the following command:

```
GRPC_TLS=true make run_addons timescale-reader timescale-writer
```


This will start the specified addons with secure communication using automatically generated certificates if they are not already available.

## Running Addons Services

To run all the addons services, you can use the following command:

```
make run_addons
```

This command will start all the addons services.

If you want to run the addons services with gRPC mTLS enabled, you can use the following command:

```
GRPC_MTLS=true make run_addons
```

If you want to run the addons services with gRPC TLS enabled, you can use the following command:

```
GRPC_TLS=true make run_addons
```

This will start all the addons services with secure communication using automatically generated certificates if they are not already available.

File: docker/addons/certs/docker-compose.yml [hunk 19]
```diff
@@ -65,12 +65,22 @@ services:
...
       - ../../ssl/certs/ca.crt:/etc/ssl/certs/ca.crt
+      - type: bind
+        source: ${MF_ADDONS_CERTS_PATH_PREFIX}${MF_USERS_GRPC_CLIENT_CERT:-./ssl/certs/dummy/client_cert}
+        target: /users-grpc-client${MF_USERS_GRPC_CLIENT_CERT:+.crt}
+      - type: bind
+        source: ${MF_ADDONS_CERTS_PATH_PREFIX}${MF_USERS_GRPC_CLIENT_KEY:-./ssl/certs/dummy/client_key}
+        target: /users-grpc-client${MF_USERS_GRPC_CLIENT_KEY:+.key}
+      - type: bind
+        source: ${MF_ADDONS_CERTS_PATH_PREFIX}${MF_USERS_GRPC_SERVER_CA_CERTS:-./ssl/certs/dummy/server_ca}
+        target: /users-grpc-server-ca${MF_USERS_GRPC_SERVER_CA_CERTS:+.crt}
```
Comment (dborovcanin) [inline docker/addons/certs/docker-compose.yml]: Add an empty line.

File: docker/addons/influxdb-reader/docker-compose.yml [hunk 20]
```diff
@@ -40,18 +40,39 @@ services:
...
+        target: /users-grpc-server-ca${MF_USERS_GRPC_SERVER_CA_CERTS:+.crt}
+      - type: bind
+        source: ${MF_ADDONS_CERTS_PATH_PREFIX}${MF_THINGS_AUTH_GRPC_CLIENT_CERT:-./ssl/certs/dummy/client_cert}
+        target: /things-grpc-client${MF_THINGS_AUTH_GRPC_CLIENT_CERT:+.crt}
+      - type: bind
+        source: ${MF_ADDONS_CERTS_PATH_PREFIX}${MF_THINGS_AUTH_GRPC_CLIENT_KEY:-./ssl/certs/dummy/client_key}
+        target: /things-grpc-client${MF_THINGS_AUTH_GRPC_CLIENT_KEY:+.key}
+      - type: bind
+        source: ${MF_ADDONS_CERTS_PATH_PREFIX}${MF_THINGS_AUTH_GRPC_SERVER_CA_CERTS:-./ssl/certs/dummy/server_ca}
+        target: /things-grpc-server-ca${MF_THINGS_AUTH_GRPC_SERVER_CA_CERTS:+.crt}
```
Comment (dborovcanin) [inline docker/addons/influxdb-reader/docker-compose.yml]: Same.

File: docker/addons/mongodb-reader/docker-compose.yml [hunk 21]
```diff
@@ -29,18 +29,39 @@ services:
...
+        target: /users-grpc-server-ca${MF_USERS_GRPC_SERVER_CA_CERTS:+.crt}
+      - type: bind
+        source: ${MF_ADDONS_CERTS_PATH_PREFIX}${MF_THINGS_AUTH_GRPC_CLIENT_CERT:-./ssl/certs/dummy/client_cert}
+        target: /things-grpc-client${MF_THINGS_AUTH_GRPC_CLIENT_CERT:+.crt}
+      - type: bind
+        source: ${MF_ADDONS_CERTS_PATH_PREFIX}${MF_THINGS_AUTH_GRPC_CLIENT_KEY:-./ssl/certs/dummy/client_key}
+        target: /things-grpc-client${MF_THINGS_AUTH_GRPC_CLIENT_KEY:+.key}
+      - type: bind
+        source: ${MF_ADDONS_CERTS_PATH_PREFIX}${MF_THINGS_AUTH_GRPC_SERVER_CA_CERTS:-./ssl/certs/dummy/server_ca}
+        target: /things-grpc-server-ca${MF_THINGS_AUTH_GRPC_SERVER_CA_CERTS:+.crt}
```
Comment (dborovcanin) [inline docker/addons/mongodb-reader/docker-compose.yml]: Same. Chack all the other files for this, as well.

File: docker/docker-compose.yml [hunk 22]
```diff
@@ -92,12 +92,15 @@ services:
...
       MF_THINGS_HTTP_PORT: ${MF_THINGS_HTTP_PORT}
-      MF_THINGS_HTTP_SERVER_CERT: ${MF_THINGS_HTTP_SERVER_CERT}
-      MF_THINGS_HTTP_SERVER_KEY: ${MF_THINGS_HTTP_SERVER_KEY}
       MF_THINGS_GRPC_HOST: ${MF_THINGS_GRPC_HOST}
       MF_THINGS_GRPC_PORT: ${MF_THINGS_GRPC_PORT}
-      MF_THINGS_GRPC_SERVER_CERT: ${MF_THINGS_GRPC_SERVER_CERT}
-      MF_THINGS_GRPC_SERVER_KEY: ${MF_THINGS_GRPC_SERVER_KEY}
+      ## Compose supports parameter expansion in environment,
+      ## Eg: ${VAR:+replacement} or ${VAR+replacement} -> replacement if VAR is set and non-empty, otherwise empty
+      ## Eg :${VAR:-default} or ${VAR-default}  -> value of VAR if set and non-empty, otherwise default
```
Comment (dborovcanin) [inline docker/docker-compose.yml:99]: I like the explanation comment.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: internal/clients/grpc/connect.go [hunk 23]
```diff
@@ -13,16 +17,25 @@ import (
 	"google.golang.org/grpc/credentials/insecure"
 )
 
+type security int
+
+const (
+	WITHOUT_TLS security = iota
```
Comment (drasko) [inline internal/clients/grpc/connect.go]: Change this to Go idiomatic way of denoting consts.

File: Makefile [hunk 24]
```diff
@@ -13,7 +13,16 @@ GOARCH ?= amd64
 VERSION ?= $(shell git describe --abbrev=0 --tags)
 COMMIT ?= $(shell git rev-parse HEAD)
 TIME ?= $(shell date +%F_%T)
-
+USER_REPO ?= $(shell git remote get-url origin | sed -e 's/.*\/\([^/]*\)\/\([^/]*\).*/\1_\2/' )
+BRANCH ?= $(shell git rev-parse --abbrev-ref HEAD  2>/dev/null || git describe --tags --abbrev=0  2>/dev/null )
+empty:=
+space:= $(empty) $(empty)
+DOCKER_PROJECT ?= $(subst $(space),,$(USER_REPO)_$(BRANCH))
```
Comment (drasko) [inline Makefile]: Can you try doing it in one line, but making it readable?

**Review by drasko — CHANGES_REQUESTED** [review]

**Review by drasko — APPROVED** [review]
LGTM

**Review by dborovcanin — APPROVED** [review]
